File: src/modules/ai-assistant-v2/provider/provider.ts
/**
 * Provider 模块 - AI模型提供商管理
 */

import type { ProviderConfig, StreamEvent } from '../types';

// 支持的提供商列表
export const SUPPORTED_PROVIDERS = [
    'openai',
    'anthropic', 
    'google',
    'deepseek',
    'ollama',
    'azure',
    'custom'
] as const;

export type ProviderType = typeof SUPPORTED_PROVIDERS[number];

// 提供商模型配置
export interface ModelInfo {
    id: string;
    name: string;
    provider: ProviderType;
    maxTokens: number;
    supportsStreaming: boolean;
    supportsTools: boolean;
    supportsVision: boolean;
}

// 默认模型配置
export const DEFAULT_MODELS: Record<ProviderType, ModelInfo[]> = {
    openai: [
        { id: 'gpt-4o', name: 'GPT-4o', provider: 'openai', maxTokens: 128000, supportsStreaming: true, supportsTools: true, supportsVision: true },
        { id: 'gpt-4o-mini', name: 'GPT-4o Mini', provider: 'openai', maxTokens: 128000, supportsStreaming: true, supportsTools: true, supportsVision: true },
        { id: 'gpt-4-turbo', name: 'GPT-4 Turbo', provider: 'openai', maxTokens: 128000, supportsStreaming: true, supportsTools: true, supportsVision: true },
    ],
    anthropic: [
        { id: 'claude-sonnet-4-20250514', name: 'Claude Sonnet 4', provider: 'anthropic', maxTokens: 200000, supportsStreaming: true, supportsTools: true, supportsVision: true },
        { id: 'claude-3-5-sonnet-20241022', name: 'Claude 3.5 Sonnet', provider: 'anthropic', maxTokens: 200000, supportsStreaming: true, supportsTools: true, supportsVision: true },
        { id: 'claude-3-opus-20240229', name: 'Claude 3 Opus', provider: 'anthropic', maxTokens: 200000, supportsStreaming: true, supportsTools: true, supportsVision: true },
    ],
    google: [
        { id: 'gemini-2.0-flash', name: 'Gemini 2.0 Flash', provider: 'google', maxTokens: 1000000, supportsStreaming: true, supportsTools: true, supportsVision: true },
        { id: 'gemini-1.5-pro', name: 'Gemini 1.5 Pro', provider: 'google', maxTokens: 2000000, supportsStreaming: true, supportsTools: true, supportsVision: true },
    ],
    deepseek: [
        { id: 'deepseek-chat', name: 'DeepSeek Chat', provider: 'deepseek', maxTokens: 64000, supportsStreaming: true, supportsTools: true, supportsVision: false },
        { id: 'deepseek-reasoner', name: 'DeepSeek Reasoner', provider: 'deepseek', maxTokens: 64000, supportsStreaming: true, supportsTools: true, supportsVision: false },
    ],
    ollama: [
        { id: 'llama3.3', name: 'Llama 3.3', provider: 'ollama', maxTokens: 128000, supportsStreaming: true, supportsTools: true, supportsVision: false },
        { id: 'qwen2.5-coder', name: 'Qwen 2.5 Coder', provider: 'ollama', maxTokens: 32000, supportsStreaming: true, supportsTools: true, supportsVision: false },
    ],
    azure: [
        { id: 'gpt-4o', name: 'Azure GPT-4o', provider: 'azure', maxTokens: 128000, supportsStreaming: true, supportsTools: true, supportsVision: true },
    ],
    custom: []
};

/**
 * Provider 类 - 管理单个AI提供商
 */
export class Provider {
    private config: ProviderConfig;
    private modelInfo: ModelInfo | null = null;

    constructor(config: ProviderConfig) {
        this.config = config;
        this.loadModelInfo();
    }

    private loadModelInfo(): void {
        const providerModels = DEFAULT_MODELS[this.config.id as ProviderType] || [];
        this.modelInfo = providerModels.find(m => m.id === this.config.model) || null;
    }

    getConfig(): ProviderConfig {
        return this.config;
    }

    getModelInfo(): ModelInfo | null {
        return this.modelInfo;
    }

    supportsStreaming(): boolean {
        return this.modelInfo?.supportsStreaming ?? true;
    }

    supportsTools(): boolean {
        return this.modelInfo?.supportsTools ?? true;
    }

    /**
     * 构建API请求头
     */
    getHeaders(): Record<string, string> {
        const headers: Record<string, string> = {
            'Content-Type': 'application/json'
        };

        switch (this.config.id) {
            case 'openai':
            case 'deepseek':
            case 'custom':
                headers['Authorization'] = `Bearer ${this.config.apiKey}`;
                break;
            case 'anthropic':
                headers['x-api-key'] = this.config.apiKey;
                headers['anthropic-version'] = '2023-06-01';
                break;
            case 'google':
                // Google使用API密钥作为查询参数
                break;
            case 'azure':
                headers['api-key'] = this.config.apiKey;
                break;
        }

        return headers;
    }

    /**
     * 获取API端点
     */
    getEndpoint(): string {
        if (this.config.baseUrl) {
            return this.config.baseUrl;
        }

        switch (this.config.id) {
            case 'openai':
                return 'https://api.openai.com/v1';
            case 'anthropic':
                return 'https://api.anthropic.com/v1';
            case 'google':
                return 'https://generativelanguage.googleapis.com/v1beta';
            case 'deepseek':
                return 'https://api.deepseek.com/v1';
            case 'ollama':
                return 'http://localhost:11434/api';
            default:
                return this.config.baseUrl || '';
        }
    }
}

/**
 * Provider 管理器
 */
export class ProviderManager {
    private providers: Map<string, Provider> = new Map();
    private currentProvider: Provider | null = null;

    /**
     * 注册提供商
     */
    register(config: ProviderConfig): Provider {
        const provider = new Provider(config);
        this.providers.set(config.id, provider);
        return provider;
    }

    /**
     * 获取提供商
     */
    get(id: string): Provider | undefined {
        return this.providers.get(id);
    }

    /**
     * 设置当前提供商
     */
    setCurrent(id: string): Provider {
        const provider = this.providers.get(id);
        if (!provider) {
            throw new Error(`Provider not found: ${id}`);
        }
        this.currentProvider = provider;
        return provider;
    }

    /**
     * 获取当前提供商
     */
    getCurrent(): Provider | null {
        return this.currentProvider;
    }

    /**
     * 列出所有提供商
     */
    list(): Provider[] {
        return Array.from(this.providers.values());
    }

    /**
     * 获取指定提供商的可用模型
     */
    getModels(providerType: ProviderType): ModelInfo[] {
        return DEFAULT_MODELS[providerType] || [];
    }

    /**
     * 获取所有支持的提供商类型
     */
    getSupportedProviders(): readonly string[] {
        return SUPPORTED_PROVIDERS;
    }
}

// 导出单例管理器
export const providerManager = new ProviderManager();

File: src-tauri/src/ai_v2/provider.rs
//! Provider 模块 - AI 模型提供商

use serde::{Deserialize, Serialize};

/// 支持的提供商类型
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProviderType {
    OpenAI,
    Anthropic,
    Google,
    DeepSeek,
    Ollama,
    Azure,
    Volcengine, // 火山引擎/豆包
    Custom,
}

impl ProviderType {
    /// 从URL检测提供商类型
    pub fn from_url(url: &str) -> Self {
        let url_lower = url.to_lowercase();
        if url_lower.contains("anthropic") || url_lower.ends_with("/v1/messages") {
            ProviderType::Anthropic
        } else if url_lower.contains("googleapis") || url_lower.contains("google") {
            ProviderType::Google
        } else if url_lower.contains("deepseek") {
            ProviderType::DeepSeek
        } else if url_lower.contains("localhost:11434") || url_lower.contains("ollama") {
            ProviderType::Ollama
        } else if url_lower.contains("azure") {
            ProviderType::Azure
        } else if url_lower.contains("volces")
            || url_lower.contains("volcengine")
            || url_lower.contains("ark.cn")
        {
            ProviderType::Volcengine
        } else if url_lower.contains("openai") {
            ProviderType::OpenAI
        } else {
            ProviderType::Custom
        }
    }

    /// 检查是否需要特殊的API格式
    pub fn uses_claude_format(&self) -> bool {
        matches!(self, ProviderType::Anthropic)
    }

    /// 获取API版本头
    pub fn api_version(&self) -> Option<&'static str> {
        match self {
            ProviderType::Anthropic => Some("2023-06-01"),
            _ => None,
        }
    }
}

/// 模型信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelInfo {
    pub id: String,
    pub name: String,
    pub provider: String,
    pub max_tokens: u32,
    pub supports_tools: bool,
    pub supports_streaming: bool,
    pub supports_vision: bool,
}

/// 默认模型列表
pub fn get_default_models() -> Vec<ModelInfo> {
    vec![
        // OpenAI
        ModelInfo {
            id: "gpt-4o".to_string(),
            name: "GPT-4o".to_string(),
            provider: "openai".to_string(),
            max_tokens: 128000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: true,
        },
        ModelInfo {
            id: "gpt-4o-mini".to_string(),
            name: "GPT-4o Mini".to_string(),
            provider: "openai".to_string(),
            max_tokens: 128000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: true,
        },
        // Anthropic
        ModelInfo {
            id: "claude-sonnet-4-20250514".to_string(),
            name: "Claude Sonnet 4".to_string(),
            provider: "anthropic".to_string(),
            max_tokens: 200000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: true,
        },
        ModelInfo {
            id: "claude-3-5-sonnet-20241022".to_string(),
            name: "Claude 3.5 Sonnet".to_string(),
            provider: "anthropic".to_string(),
            max_tokens: 200000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: true,
        },
        // DeepSeek
        ModelInfo {
            id: "deepseek-chat".to_string(),
            name: "DeepSeek Chat".to_string(),
            provider: "deepseek".to_string(),
            max_tokens: 64000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: false,
        },
        ModelInfo {
            id: "deepseek-reasoner".to_string(),
            name: "DeepSeek Reasoner".to_string(),
            provider: "deepseek".to_string(),
            max_tokens: 64000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: false,
        },
        // Google
        ModelInfo {
            id: "gemini-2.0-flash".to_string(),
            name: "Gemini 2.0 Flash".to_string(),
            provider: "google".to_string(),
            max_tokens: 1000000,
            supports_tools: true,
            supports_streaming: true,
            supports_vision: true,
        },
    ]
}

/// 获取API请求头
pub fn get_headers(provider: ProviderType, api_key: &str) -> Vec<(String, String)> {
    let mut headers = vec![("Content-Type".to_string(), "application/json".to_string())];

    match provider {
        ProviderType::Anthropic => {
            headers.push(("x-api-key".to_string(), api_key.to_string()));
            headers.push(("anthropic-version".to_string(), "2023-06-01".to_string()));
        }
        ProviderType::Azure => {
            headers.push(("api-key".to_string(), api_key.to_string()));
        }
        _ => {
            headers.push(("Authorization".to_string(), format!("Bearer {}", api_key)));
        }
    }

    headers
}

/// 构建API端点
pub fn build_endpoint(base_url: &str, provider: ProviderType) -> String {
    let base = base_url.trim_end_matches('/');
    let base_lower = base.to_lowercase();

    if base_lower.ends_with("/chat/completions")
        || base_lower.ends_with("/v1/messages")
        || base_lower.ends_with("/api/chat")
    {
        return base.to_string();
    }

    match provider {
        ProviderType::Anthropic => {
            if base_lower.ends_with("/v1") {
                format!("{}/messages", base)
            } else {
                format!("{}/v1/messages", base)
            }
        }
        ProviderType::Ollama => {
            if base_lower.ends_with("/v1") {
                format!("{}/chat/completions", base)
            } else {
                format!("{}/v1/chat/completions", base)
            }
        }
        // 火山引擎API - base_url已经包含完整路径，直接添加chat/completions
        ProviderType::Volcengine => format!("{}/chat/completions", base),
        _ => format!("{}/chat/completions", base),
    }
}
